File: src/main/java/com/cloudstorage/security/JwtService.java
package com.cloudstorage.security;

import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.security.Key;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * Handles JWT creation and validation.
 *
 * The signing key is derived from the JWT_SECRET environment variable.
 * The secret must be at least 32 characters (256 bits) for HS256.
 * NEVER hardcode the secret in source code.
 */
@Service
public class JwtService {

    private static final Logger log = LoggerFactory.getLogger(JwtService.class);

    @Value("${app.jwt.secret}")
    private String jwtSecret;

    @Value("${app.jwt.expiration-ms}")
    private long jwtExpirationMs;

    // ── Token Generation ──────────────────────────────────────────────────────

    /**
     * Generates a signed JWT containing the user's ID, email, and role.
     */
    public String generateToken(String userId, String email, String role) {
        Map<String, Object> claims = new HashMap<>();
        claims.put("userId", userId);
        claims.put("role",   role);

        return Jwts.builder()
                .setClaims(claims)
                .setSubject(email)
                .setIssuedAt(new Date())
                .setExpiration(new Date(System.currentTimeMillis() + jwtExpirationMs))
                .signWith(getSigningKey(), SignatureAlgorithm.HS256)
                .compact();
    }

    // ── Claim Extraction ──────────────────────────────────────────────────────

    public String extractEmail(String token) {
        return extractClaim(token, Claims::getSubject);
    }

    public String extractUserId(String token) {
        return extractClaim(token, c -> c.get("userId", String.class));
    }

    public String extractRole(String token) {
        return extractClaim(token, c -> c.get("role", String.class));
    }

    // ── Validation ────────────────────────────────────────────────────────────

    public boolean isTokenValid(String token) {
        try {
            Jwts.parserBuilder()
                    .setSigningKey(getSigningKey())
                    .build()
                    .parseClaimsJws(token);
            return true;
        } catch (ExpiredJwtException e) {
            log.warn("JWT expired: {}", e.getMessage());
        } catch (UnsupportedJwtException e) {
            log.warn("JWT unsupported: {}", e.getMessage());
        } catch (MalformedJwtException e) {
            log.warn("JWT malformed: {}", e.getMessage());
        } catch (SecurityException e) {
            log.warn("JWT signature invalid: {}", e.getMessage());
        } catch (IllegalArgumentException e) {
            log.warn("JWT claims empty: {}", e.getMessage());
        }
        return false;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private <T> T extractClaim(String token, Function<Claims, T> resolver) {
        return resolver.apply(extractAllClaims(token));
    }

    private Claims extractAllClaims(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(getSigningKey())
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    private Key getSigningKey() {
        // Use raw bytes so the secret can be a plain string (must be ≥ 32 chars for HS256)
        return Keys.hmacShaKeyFor(jwtSecret.getBytes(StandardCharsets.UTF_8));
    }
}

File: src/services/jwt.service.js
// ── JWT Service ───────────────────────────────────────────────────────────────
// Signs and verifies JWTs using HS256. 
// Claims match the original Java JwtService exactly:
//   - subject = email
//   - claims: userId, role

const jwt = require('jsonwebtoken');

const JWT_SECRET     = process.env.JWT_SECRET     || 'cloudvault-dev-secret-key-minimum-32-characters-long';
const JWT_EXPIRES_IN = process.env.JWT_EXPIRATION_MS
  ? Math.floor(Number(process.env.JWT_EXPIRATION_MS) / 1000) + 's'
  : '86400s'; // 24 hours

/**
 * Generates a signed JWT token.
 */
function generateToken(userId, email, role) {
  return jwt.sign(
    { userId, role },      // payload claims
    JWT_SECRET,
    {
      subject:   email,    // sub claim = email (matches Java)
      algorithm: 'HS256',
      expiresIn: JWT_EXPIRES_IN,
    }
  );
}

/**
 * Verifies a token and returns the decoded payload.
 * Throws if the token is invalid or expired.
 */
function verifyToken(token) {
  return jwt.verify(token, JWT_SECRET, { algorithms: ['HS256'] });
}

module.exports = { generateToken, verifyToken };
